rfcs/0039-unprivileged-maintainer-teams.md

```diff
@@ -0,0 +1,176 @@
+change the maintainer list to include the GitHub user *ID* as well as
+their handle. When syncing, the automation will validate the user ID
+matches. GitHub User IDs are easily found at
+`https://api.github.com/users/«username»`.
+
+If a user ID's GitHub handle changes, the maintainer should remain
+part of the team under their new name. The repository's nickname
+should be updated to reflect their new nickname.
+
+## Team Automation
+
+The team must be automatically updated at least once a day to ensure
+the maintainer list is fresh and up to date. The automation for this
+will be written in Rust with the hubcaps library. It will run on the
+NixOS infrastructure with limited credentials, with only sufficient
+permission to manage the team.
+
+The automation will fetch a fresh version of Nixpkgs, extract the
+maintainer information, and update the team. It will support a dry-run
+option.
+
+New members of the team will receive an invitation to join the GitHub
+organization.
+
+## OfBorg changes
+
+OfBorg will identify PRs which are approved by their maintainers, and
+add a special label `approved-by-maintainer`.
+
+## Roll-Out Plan
+
+1. Write an explanatory post on Discourse about the what-and-why of
+   this plan.
+2. Select a small group of maintainers who are not committers to be
+   part of the first round, and manually run the tooling, and pause
+   half a week to see what changes.
+3. Automate the tooling on the infrastructure.
+4. Expand the group to one quarter of the maintainers, and pause a
+   half a week to gauge response.
+5. Expand the group to one half of the maintainers and wait one week.
+6. Expand the group to all of the maintainers.
+
+If we receive no major feedback or problems during the rollout, we
+will continue to 100%.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ - Putting each maintainer in a read only team will display
+   maintainers as "member", without specifying which team they are a
+   member of. This gives the impression of authority which maintainers
+   don't already receive. This is a pro and a con.
+
+ - A mistake in the automation, or in the admin panel of GitHub could
+   grant the team write access to Nix ecosystem repositories.
+
+ - Package maintainers who do not wish to have a GitHub account will
+   not benefit from this change.
+
+ - Package maintainers who do have a GitHub account, but do not wish
+   to use 2 factor authentication will not benefit from this change.
+
+ - If a GitHub user is a real jerk on the internet, it can potentially
+   reflect negatively on the NixOS ecosystem. Someone who is banned
+   from the NixOS GitHub organization is not allowed to be a package
+   maintainer.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Mentioning people in GitHub comments is the main alternative. This has
+the major down-side of not receiving the support of [GitHub's UI
+for requested reviews](https://github.com/pulls/review-requested).
```

Comment:
How important is the GitHub UI for requesting reviews? Addresses most of the drawbacks, and I'm not sure the UI is that valuable in comparison.

Reply:
Creating the team seems fairly cheap to me.  /cc comments will create unnecessary noise every time, similarly to what borg did before being moved to the "checks" tab.

Reply:
Another alternative, not mentioned here, is to just automatically send mail to maintainers. This would require a custom program to be written, but so would the proposal here. No GitHub features required.

Reply:
Additionally, it'd work for people who are not using GitHub. The only drawback (compared to this RFC) would be that reviews from maintainers wouldn't show up in color, but in gray -- but if ofborg adds `approved-by-maintainer` anyway that's likely not an issue ; esp. as otherwise reviews from maintainers of one package would appear in color for other packages too.

Reply:
Is it realistically bearable to be an active Nixpkgs package maintainer without Github account? I don't see many patches posted to mailing list, at least.

Reply:
I think requesting reviews is much better than a `cc` comment. The expectation is the requestee will review it. A `cc` comment doesn't convey such an expectation. Additionally, by using the tooling specific for this purpose, we get for free all the workflow improvements GitHub provides.  I don't think it is worth optimizing for the few users who doesn't want to have a GitHub account, while sacrificing the user experience of the many who do.

Reply:
You could add another Check that sends emails and reports back to the PR but it wouldn't be as prominent as adding the reviewers to the list. It's quite nice when submitting a PR to know who is going to review your work.

Reply:
We already use review requests as more of a notification mechanism than assigning work to somebody. So, I think that showing review requests is at best not useful and at worst misleading.

Reply:
I think we are on the path of changing behavior, and making better use of our tools. I don't think we should get too bogged down in how we currently misuse features when considering how we can make better use later.

Reply:
> we get for free all the workflow improvements GitHub provides.

Can you explain what these are? I'm not sure that's been addressed so far in this RFC.

Reply:
a note about why we should use review requests:

This takes advantage of a structured way to express a review is requested. This data has specific meaning to GitHub and all API clients of GitHub. By declaring yourself a maintainer, you have effectively taken ownership of the thing being changed. Changing something owned by someone else, ultimately (in the future where maintainership in Nixpkgs means more than it has in the past) should get a review by the owner. Taking advantage of the semantically reasonable mechanism to do this means GitHub can make improvements based on what they know that data to mean.

---

Specific examples include the linked page listing PRs where your review is requested (but you haven't replied yet). The future workflows they could build are unknowable. There are also third-party dashboards people use (though I can't remember their names :|) which display review requests differently.

Reply:
I think we shouldn't use review requests for notifications. We should use `cc` comments for that.

Reply:
Fair enough.

Reply:
While a page listing PRs where one's review is requested without answer yet is an argument, I don't think we should choose our workflow based on hypothetical improvements GitHub could do later on. First, make the tooling, then decide the workflow accordingly.

Now, whether a page listing PRs where the review is requested and/or third-party dashboards are an advantage enough to warrant this change is another question. I personally don't use them and would assume people who maintain just one or two packages would see even less use to these, but may be wrong here :)

Reply:
I think given:
 (a) various interesting corner cases in Github behaviour we hit from time to time,
 (b) our story of migrating basically everything exchanging one featureset to another one
we shouldn't really look too much at what Github _intends_ things to mean and ask which set of the side-effects we like better.

(I do probably like side-effects of review requests better than of a comment, unless there is a unified comment with everything ofBorg wants to say about the PR)

Separately, I think that making maintainership mean much more (even if desirable) has time horizon larger than replacing at least one part of the tool collection; so discussing how multiple tools (Github, ofborg, this team maintenance bot) interact can assume nothing big changes in the process for low-impact packages.

Reply:
I agree we shouldn't bank on GitHub making big changes. I do think the existing tooling is very good already. I think changing behavior is much faster than you say -- small nudges have far reduced direct pushes within a short period of time of having ofborg do reasonable checks and validation.

Reply:
I think that works for tooling adoption — and high standards you uphold for ofborg help — better than for shifting general notions (especially in the direction that creates more single-person-of-delay situations).

Anyway, I agree that in the current situation review requests make the most sense.

Reply:
Having thought about this more, this isn’t actually mutually exclusive with an option that did not depend on GitHub, especially since implementing an email-based system would require no priveleges whatsoever, and could be done as an opt-in system by any community member. While I’m among the (probable) minority of community members who would like Nixpkgs development to move away from GitHub, this doesn’t actually increase our lock-in at all, and would be strictly better than what we have now.